search/mongo_search.go

```diff
@@ -359,8 +329,136 @@ func (m *MongoSearcher) createParamObjects(params []SearchParam) []bson.M {
 }
 
 func (m *MongoSearcher) createPipelineObject(query Query) []bson.M {
-	// Currently this is just a $match stage with the same query that would be used in find()
-	return []bson.M{{"$match": m.createQueryObject(query)}}
+	standardSearchParams := []SearchParam{}
+	chainedSearchParams := []SearchParam{}
+
+	// Separate out chained search parameters
+	for _, p := range query.Params() {
+		if usesChainedSearch(p) {
+			chainedSearchParams = append(chainedSearchParams, p)
+			continue
+		}
+		standardSearchParams = append(standardSearchParams, p)
+	}
+
+	// Process standard SearchParams
+	pipeline := []bson.M{{"$match": m.createQueryObjectFromParams(standardSearchParams)}}
+
+	// Process chained search parameters
+	for _, p := range chainedSearchParams {
+		pipeline = append(pipeline, m.createChainedSearchPipelineStages(p)...)
+	}
+
+	return pipeline
+}
+
+// The SearchParam argument should be either a ReferenceParam or an OrParam.
+func (m *MongoSearcher) createChainedSearchPipelineStages(searchParam SearchParam) []bson.M {
+	// This returns 2 stages in the pipeline that represent a chained query reference:
+	// 1. A $lookup for the foreign Resource being referenced
+	// 2. A $match on that foreign Resource
+	stages := make([]bson.M, 2)
+
+	// Build the $lookup. We need to get a ReferenceParam (of type ChainedQueryReference)
+	// that we can use to populate the $lookup. If it's an OR, any one of its Items
+	// should do.
+	var lookupRef *ReferenceParam
+	var ok bool
+
+	_, isOr := searchParam.(*OrParam)
+
+	if isOr {
+		lookupRef, ok = searchParam.(*OrParam).Items[0].(*ReferenceParam)
+		if !ok {
+			panic(createInternalServerError("", ""))
+		}
+	} else {
+		lookupRef = searchParam.(*ReferenceParam)
+	}
+
+	chainedRef, ok := lookupRef.Reference.(ChainedQueryReference)
+	if !ok {
+		panic(createInternalServerError("", ""))
+	}
+
+	collectionName := models.PluralizeLowerResourceName(chainedRef.Type)
+
+	stages[0] = bson.M{"$lookup": bson.M{
+		"from":         collectionName,
+		"localField":   convertSearchPathToMongoField(lookupRef.getInfo().Paths[0].Path) + ".referenceid",
+		"foreignField": "_id",
+		"as":           "_" + collectionName,
```

Comment:
While naming the "as" after the collection name works in many cases, it will cause bugs when multiple chains are used that reference the same type of collection.  How often is that likely to occur?  I don't know -- but you can avoid the trouble by naming it after the search param instead...  BTW -- if you are going to support multiple paths (as noted above), then the as might be the search param with the index of the path appended.